var/spack/repos/builtin/packages/plasma/package.py

```diff
@@ -58,36 +61,63 @@ class Plasma(MakefilePackage):
     conflicts("%xl_r")
 
     patch("remove_absolute_mkl_include.patch", when="@17.1")
-    patch("add_netlib_lapacke_detection.patch", when="@17.1")
-
-    def getblaslapacklibs(self):
-        if '^netlib-lapack' in self.spec:
-            bl_attr = ':c,fortran'
-        else:
-            bl_attr = ''
-        return self.spec['lapack' + bl_attr].libs + \
-            self.spec['blas' + bl_attr].libs + \
-            find_system_libraries(['libm'])
 
+    @when("@18.9.0:")
+    def cmake_args(self):
+        if self.spec.satisfies('@:17.1'):
+            open("CMakeLists.txt", "w").write("cmake_minimum_required( VERSION 3.0 FATAL_ERROR )\n")
+
+        options = list()
+
+        options.extend([
+            "-DCMAKE_INSTALL_PREFIX=%s" % prefix,
+            "-DCMAKE_INSTALL_NAME_DIR:PATH=%s/lib" % prefix,
+            "-DBLAS_LIBRARIES=%s" % self.spec["blas"].libs.joined(";"),
+            "-DLAPACK_LIBRARIES=%s" % self.spec["lapack"].libs.joined(";")
+        ])
+
+        options += [
+            "-DBUILD_SHARED_LIBS=%s" %
+            ('ON' if ('+shared' in self.spec) else 'OFF')
+        ]
+
+        return options
+
+    # Before 18.9.0 it was an Makefile package
+    @when("@:17.1")
+    def cmake(self, spec, prefix):
+        pass
+
+    # Before 18.9.0 it was an Makefile package
+    @when("@:17.1")
+    def build(self, spec, prefix):
+        pass
+
+    # Before 18.9.0 it was an Makefile package
+    @when("@:17.1")
+    def install(self, spec, prefix):
+        self.edit(spec, prefix)
+        make()
+        make("install")
+
+    @when("@:17.1")
     def edit(self, spec, prefix):
         # copy "make.inc.mkl-gcc" provided by default into "make.inc"
         open("make.inc", "w").write(open("make.inc.mkl-gcc").read())
 
         make_inc = FileFilter("make.inc")
 
-        if '~shared' in self.spec:
-            make_inc.filter("-fPIC", "")  # not using fPIC
-
-        if "^mkl" not in spec:
+        if not spec.satisfies("^intel-mkl"): # or not env.has_key("MKLROOT"):
             make_inc.filter("-DPLASMA_WITH_MKL", "")  # not using MKL
-            make_inc.filter("-DHAVE_MKL", "")         # not using MKL
+            make_inc.filter("LIBS *= *.*", "LIBS = " +
+                self.spec["blas"].libs.ld_flags + " -lm")
 
         header_flags = ""
         # accumulate CPP flags for headers: <cblas.h> and <lapacke.h>
         for dep in ("blas", "lapack"):
             try:  # in case the dependency does not provide header flags
```

Comment:
but how would it find the headers then? If it doesn't work (or we might pick up system libraries) just throw an error and tell the user how to fix it.

Reply:
This is what `for dep` loop is doing: walk dependencies, get the header paths, add it to build. Dependencies that do not provide headers are explicitly blocked and the comments next to `conflict()` say that the headers are missing or incomplete.

Reply:
So we may never hit the `except` block because we explicitly disable all non-complete providers? Why do we ignore the exception then? What am I missing?

Suppose someone adds a new provider that also doesn't provide all headers.

Reply:
The reason why there is a `try-except` because I had an exception for one of the combinations of dependencies.

Providing "blas" and "lapack" is not a well defined functionality. Some of this is taken care of by conflict designation. The usable packages are handled here. But the problem is "blas" might not provide something because "lapack" provides that. And looking at the packages of "blas" and "lapack", I didn't see a common theme I could just use in a consistent manner.